src/recordlinker/linking/link.py

```diff
@@ -139,43 +140,46 @@ def link_record_against_mpi(
     for linkage_pass in algo_config:
-        # the minimum ratio of matches needed to be considered a cluster member
-        cluster_ratio = linkage_pass.get("cluster_ratio", 0)
-        # initialize a dictionary to hold the clusters of patients for each person
-        clusters: dict[models.Person, list[models.Patient]] = collections.defaultdict(
-            list
-        )
-        # block on the pii_record and the algorithm's blocking criteria, then
-        # iterate over the patients, grouping them by person
-        patients = mpi_service.get_block_data(session, pii_record, linkage_pass)
-        for patient in patients:
-            clusters[patient.person].append(patient)
-
-        # evaluate each Person cluster to see if the incoming record is a match
-        for person, patients in clusters.items():
-            matched_count = 0
-            for patient in patients:
-                # increment our match count if the pii_record matches the patient
-                if compare(pii_record, patient, linkage_pass):
-                    matched_count += 1
-            # calculate the match ratio for this person cluster
-            match_ratio = matched_count / len(patients)
-            if match_ratio >= cluster_ratio:
-                # The match ratio is larger than the minimum cluster threshold,
-                # optionally update the max score for this person
-                scores[person] = max(scores[person], match_ratio)
+        with TRACER.start_as_current_span("link.pass"):
+            # the minimum ratio of matches needed to be considered a cluster member
+            cluster_ratio = linkage_pass.get("cluster_ratio", 0)
+            # initialize a dictionary to hold the clusters of patients for each person
+            clusters: dict[models.Person, list[models.Patient]] = collections.defaultdict(list)
+            # block on the pii_record and the algorithm's blocking criteria, then
+            # iterate over the patients, grouping them by person
+            with TRACER.start_as_current_span("link.block"):
+                patients = mpi_service.get_block_data(session, pii_record, linkage_pass)
+                for patient in patients:
+                    clusters[patient.person].append(patient)
+
+            # evaluate each Person cluster to see if the incoming record is a match
+            with TRACER.start_as_current_span("link.evaluate"):
+                for person, patients in clusters.items():
+                    matched_count = 0
+                    for patient in patients:
+                        # increment our match count if the pii_record matches the patient
+                        with TRACER.start_as_current_span("link.compare"):
+                            if compare(pii_record, patient, linkage_pass):
+                                matched_count += 1
+                    # calculate the match ratio for this person cluster
+                    match_ratio = matched_count / len(patients)
```

Comment:
I don't think it would ever happen but could `len(patients)` ever be 0? I'm thinking no since you need at least one patient for a cluster. But just thought of that edge case since its dividing by length here

Reply:
It's a good thought.  It's not possible now, at least from what I can read of the code.  If you look at this [block where the clusters are created](https://github.com/CDCgov/RecordLinker/blob/main/src/recordlinker/linking/link.py#L145-L152), a Person can only be created if a Patient record already exists.  Therefore, at a minimum, the length has to be 1.

However, there is a chance in the future that someone changes this block, which introduces a case where it could be of length 0, in which case we'd get this weird ZeroDivisionError.  Probably safest to put an assertion at the top of this loop to make sure the length isn't 0.  A custom error message will help debug that if the clustering logic ever gets changed.

Reply:
added. https://github.com/CDCgov/RecordLinker/pull/66/commits/2ee6d5ec9feff4b78e31bfbd0fe845efc4953767